In [1]:
import pandas as pd

# 아래 실습을 위해서는 Local(개인 PC)에 MySQL 서버가 설치되어 있어야 합니다.

# 1. DataFrame to DB

## 데이터 가져오기

In [2]:
# ! pip install lxml

* 먼저 데이터를 준비해봅시다. 이전 시간에 배운 API 를 이용해 데이터를 가져올 거예요.

In [5]:
import requests

serviceKey = ""  # 일반 인증키 (Decoding)

url = 'http://apis.data.go.kr/B551182/yadmOpCloInfoService1/getHospPharmacyOpCloList'
params ={'serviceKey' : serviceKey, 'numOfRows' : '10', 'pageNo' : '1', 'crtrYm' : '202401', 'yadmTp' : '1', 'opCloTp' : '1' }

response = requests.get(url, params=params)
# print(response.content)  # bytes 형태여서 보기 힘듦.
content = response.content.decode('utf-8')  # utf-8로 디코딩함. 
print(content)

<?xml version="1.0" encoding="UTF-8" standalone="yes"?><response><header><resultCode>00</resultCode><resultMsg>NORMAL SERVICE.</resultMsg></header><body><items><item><addr>인천광역시 계양구 계양대로 85, 2층 (작전동)</addr><clCdNm>한의원</clCdNm><crtrYm>202401</crtrYm><estbCnclTp>개업</estbCnclTp><estbDd>20240103</estbDd><telno>032-507-1075</telno><yadmNm>조각목한의원</yadmNm><ykiho>JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxMzUxIzUxIyQxIyQxIyQ3MiQyNjEyMjIjODEjJDEjJDYjJDgz</ykiho></item><item><addr>서울특별시 구로구 경인로15길 4, SC메디칼센터 4층 401호 (오류동)</addr><clCdNm>치과의원</clCdNm><crtrYm>202401</crtrYm><estbCnclTp>개업</estbCnclTp><estbDd>20240103</estbDd><telno>02-532-3423</telno><yadmNm>길치과의원</yadmNm><ykiho>JDQ4MTg4MSM1MSMkMSMkMCMkNzIkMzgxOTYxIzMxIyQxIyQzIyQ3OSQyNjE0ODEjNzEjJDEjJDgjJDgz</ykiho></item><item><addr>경상북도 구미시 금오대로 307-3, 1층 (오태동)</addr><clCdNm>한의원</clCdNm><crtrYm>202401</crtrYm><estbCnclTp>개업</estbCnclTp><estbDd>20240124</estbDd><telno>054-461-2728</telno><yadmNm>수평한의원</yadmNm><ykiho>JDQ4MTYyMiM4MSMkMSMkMCMkMTMkMzgxNzAyIzIxIy

* 가져온 데이터를 Data Frame 으로 변경해봅시다. 

In [6]:
df = pd.read_xml(response.content, xpath=".//item")  # 추가된 코드
df

addr clCdNm  crtrYm  \
0                        인천광역시 계양구 계양대로 85, 2층 (작전동)    한의원  202401   
1          서울특별시 구로구 경인로15길 4, SC메디칼센터 4층 401호 (오류동)   치과의원  202401   
2                      경상북도 구미시 금오대로 307-3, 1층 (오태동)    한의원  202401   
3                    대전광역시 동구 동서대로1705번길 5, 2층 (용전동)    한의원  202401   
4  서울특별시 중랑구 용마산로 670, B동 2층 206호 (망우동, 신내역 프라디움 ...    한의원  202401   
5             경기도 부천시 원미구 수도로 202, 은성프라자 2층일부호 (도당동)     의원  202401   
6             경기도 안양시 동안구 경수대로 555, 개성하이뷰 207호 (호계동)     의원  202401   
7                         강원특별자치도 원주시 북원로 2245 (단계동)     의원  202401   
8                      경기도 하남시 감일백제로 181, 403호 (감이동)    한의원  202401   
9                             전라남도 담양군 금성면 덕성길 11-58     의원  202401   

  estbCnclTp    estbDd          telno                yadmNm  \
0         개업  20240103   032-507-1075                조각목한의원   
1         개업  20240103    02-532-3423                 길치과의원   
2         개업  20240124   054-461-2728                 수평한의원   
3         개업  20240109  010-1577-2721               소리닥터한의원   
4         개업  20240126    02-454-9998               신내경희한의원   
5         개업  20240105   032-657-0119  건강한세상의료소비자생활협동조합고운의원   
6         개업  20240123   031-391-7277                 밀라비의원   
7         개업  20240112   033-743-9344               이상진안과의원   
8         개업  20240115   02-6958-7582             조민경 경희한의원   
9         개업  20240130   061-383-7600                 차오름의원   

                                               ykiho shwSbjtCdNm  
0  JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxMzUxIzUxIyQxIy...        None  
1  JDQ4MTg4MSM1MSMkMSMkMCMkNzIkMzgxOTYxIzMxIyQxIy...        None  
2  JDQ4MTYyMiM4MSMkMSMkMCMkMTMkMzgxNzAyIzIxIyQxIy...        None  
3  JDQ4MTYyMiM2MSMkMSMkMiMkMTMkMzgxNzAyIzExIyQxIy...        None  
4  JDQ4MTg4MSM1MSMkMSMkMCMkMTMkNDgxMTkxIzExIyQxIy...        None  
5  JDU4MTI3MSM1MSMkMiMkMCMkMDAkNDgxMzUxIzIxIyQxIy...         일반의  
6  JDU4MTI3MSM1MSMkMiMkMCMkMDAkNDgxMzUxIzIxIyQxIy...         일반의  
7  JDQ4MTYyMiM1MSMkMiMkNCMkMDAkMzgxMzUxIzMxIyQxIy...          안과  
8  JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxNzAyIzIxIyQxIy...        None  
9  JDQ4MTYyMiM3MSMkMiMkMCMkMDAkMzgxMTkxIzExIyQxIy...         일반의

## Pymysql을 이용하기

* Pymysql의 경우 DB 연결만 지원해주기 때문에 DB 연결을 직접 관리하고, SQL 쿼리를 수동으로 작성해서 실행해야합니다.

### 1. PyMySQL 라이브러리 import 

In [7]:
import pymysql

### 2. connect 메소드를 이용해 DB에 연결.

In [8]:
# 데이터베이스 연결 설정
connection = pymysql.connect(host='localhost',  # 데이터베이스 서버 주소
                             user='root',  # 해당 DB에서 사용하는 user 이름
                             password='0000',  # 해당 DB의 해당 user의 비밀번호
                             database='my_test',  # 접속하려는 DB명
                             charset='utf8mb4',  # 가장 기본적인 charset.
                             cursorclass=pymysql.cursors.DictCursor)

### 3. cursor 객체 가져오기. 

In [9]:
cursor = connection.cursor()  # DB 내 위치.
# 위 코드 실행 후 영향을 받은 행의 수를 반환함. DDL의 경우 영향을 받은 행은 없기 때문에 0 반환. 

### 4. cursor의 execute() 메소드를 사용해 SQL문을 DB 서버에 보냄.

In [10]:
# DB가 존재하지 않기 때문에 테이블 생성
create_table_query = """
CREATE TABLE IF NOT EXISTS hospital_pymysql (
    addr VARCHAR(255),
    clCdNm VARCHAR(255),
    crtrYm VARCHAR(255),
    estbCnclTp VARCHAR(255),    
    estbDd VARCHAR(255),
    telno VARCHAR(255),
    yadmNm VARCHAR(255),    
    ykiho VARCHAR(255),    
    shwSbjtCdNm VARCHAR(255)     
);
"""
cursor.execute(create_table_query)


0

### 5. DML의 경우 commit()을 사용해 데이터 갱신 확정.

In [11]:
# DataFrame의 각 행을 데이터베이스에 삽입

insert_query = """
INSERT INTO hospital_pymysql 
VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s);
"""
# 여기서 %s는 SQL 에서 쿼리를 작성할 때 사용되는 파라미터의 플레이스 홀더. 
# 쿼리가 실행될 때 치환될 값의 위치를 나타냄.

# DataFrame에 저장된 데이터를 row-by-row(행별로) 데이터베이스에 삽입
for _, row in df.iterrows():
    cursor.execute(insert_query, (row['addr'], row['clCdNm'], row['crtrYm'], row['estbCnclTp'], 
                                  row['estbDd'], row['telno'], row['yadmNm'], row['ykiho'], row['shwSbjtCdNm']))

# 변경 사항을 데이터베이스에 반영. DML(Insert/Update/Delete)의 경우 commit을 사용해 데이터 확정 갱신. 
connection.commit()

실제 실행되는 쿼리 한 줄 예시:  
"INSERT INTO hospital_pymysql VALUES ("인천광역시 계양구 계양대로 85, 2층 (작전동)", "한의원", "202401", "개업", "20240103", "032-507-1075", "조각목한의원", "JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxMzUxIzUxIyQxIy", "");"

### 6. 데이터 가져오기

In [12]:
cursor.execute("Select * From hospital_pymysql;")

10

In [13]:
# Fetch를 이용해 DB에서 검색된 결과를 실제로 가져옴.
cursor.fetchall()  # fetchall(), fetchone(), fetchmany() 사용 가능.

[{'addr': '인천광역시 계양구 계양대로 85, 2층 (작전동)',
  'clCdNm': '한의원',
  'crtrYm': '202401',
  'estbCnclTp': '개업',
  'estbDd': '20240103',
  'telno': '032-507-1075',
  'yadmNm': '조각목한의원',
  'ykiho': 'JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxMzUxIzUxIyQxIyQxIyQ3MiQyNjEyMjIjODEjJDEjJDYjJDgz',
  'shwSbjtCdNm': None},
 {'addr': '서울특별시 구로구 경인로15길 4, SC메디칼센터 4층 401호 (오류동)',
  'clCdNm': '치과의원',
  'crtrYm': '202401',
  'estbCnclTp': '개업',
  'estbDd': '20240103',
  'telno': '02-532-3423',
  'yadmNm': '길치과의원',
  'ykiho': 'JDQ4MTg4MSM1MSMkMSMkMCMkNzIkMzgxOTYxIzMxIyQxIyQzIyQ3OSQyNjE0ODEjNzEjJDEjJDgjJDgz',
  'shwSbjtCdNm': None},
 {'addr': '경상북도 구미시 금오대로 307-3, 1층 (오태동)',
  'clCdNm': '한의원',
  'crtrYm': '202401',
  'estbCnclTp': '개업',
  'estbDd': '20240124',
  'telno': '054-461-2728',
  'yadmNm': '수평한의원',
  'ykiho': 'JDQ4MTYyMiM4MSMkMSMkMCMkMTMkMzgxNzAyIzIxIyQxIyQ5IyQ5MiQzNjE0ODEjNDEjJDEjJDgjJDgz',
  'shwSbjtCdNm': None},
 {'addr': '대전광역시 동구 동서대로1705번길 5, 2층 (용전동)',
  'clCdNm': '한의원',
  'crtrYm': '202401',
  'estbCnclT

### 7. 가져온 데이터를 DataFrame으로 만들기

In [14]:
cursor.execute("Select * From hospital_pymysql;")
selected_date = cursor.fetchall()

In [15]:
pd.DataFrame(selected_date)

addr clCdNm  crtrYm  \
0                        인천광역시 계양구 계양대로 85, 2층 (작전동)    한의원  202401   
1          서울특별시 구로구 경인로15길 4, SC메디칼센터 4층 401호 (오류동)   치과의원  202401   
2                      경상북도 구미시 금오대로 307-3, 1층 (오태동)    한의원  202401   
3                    대전광역시 동구 동서대로1705번길 5, 2층 (용전동)    한의원  202401   
4  서울특별시 중랑구 용마산로 670, B동 2층 206호 (망우동, 신내역 프라디움 ...    한의원  202401   
5             경기도 부천시 원미구 수도로 202, 은성프라자 2층일부호 (도당동)     의원  202401   
6             경기도 안양시 동안구 경수대로 555, 개성하이뷰 207호 (호계동)     의원  202401   
7                         강원특별자치도 원주시 북원로 2245 (단계동)     의원  202401   
8                      경기도 하남시 감일백제로 181, 403호 (감이동)    한의원  202401   
9                             전라남도 담양군 금성면 덕성길 11-58     의원  202401   

  estbCnclTp    estbDd          telno                yadmNm  \
0         개업  20240103   032-507-1075                조각목한의원   
1         개업  20240103    02-532-3423                 길치과의원   
2         개업  20240124   054-461-2728                 수평한의원   
3         개업  20240109  010-1577-2721               소리닥터한의원   
4         개업  20240126    02-454-9998               신내경희한의원   
5         개업  20240105   032-657-0119  건강한세상의료소비자생활협동조합고운의원   
6         개업  20240123   031-391-7277                 밀라비의원   
7         개업  20240112   033-743-9344               이상진안과의원   
8         개업  20240115   02-6958-7582             조민경 경희한의원   
9         개업  20240130   061-383-7600                 차오름의원   

                                               ykiho shwSbjtCdNm  
0  JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxMzUxIzUxIyQxIy...        None  
1  JDQ4MTg4MSM1MSMkMSMkMCMkNzIkMzgxOTYxIzMxIyQxIy...        None  
2  JDQ4MTYyMiM4MSMkMSMkMCMkMTMkMzgxNzAyIzIxIyQxIy...        None  
3  JDQ4MTYyMiM2MSMkMSMkMiMkMTMkMzgxNzAyIzExIyQxIy...        None  
4  JDQ4MTg4MSM1MSMkMSMkMCMkMTMkNDgxMTkxIzExIyQxIy...        None  
5  JDU4MTI3MSM1MSMkMiMkMCMkMDAkNDgxMzUxIzIxIyQxIy...         일반의  
6  JDU4MTI3MSM1MSMkMiMkMCMkMDAkNDgxMzUxIzIxIyQxIy...         일반의  
7  JDQ4MTYyMiM1MSMkMiMkNCMkMDAkMzgxMzUxIzMxIyQxIy...          안과  
8  JDQ4MTYyMiM1MSMkMSMkMCMkMTMkNDgxNzAyIzIxIyQxIy...        None  
9  JDQ4MTYyMiM3MSMkMiMkMCMkMDAkMzgxMTkxIzExIyQxIy...         일반의

### cursor 및 connection 종료

In [16]:
cursor.close()
connection.close()

## ORM 라이브러리인 SQLAlchemy 사용

In [17]:
import pandas as pd
from sqlalchemy import create_engine

# MySQL 연결 설정
username = 'root'
password = '0000'
host = 'localhost'  # 데이터베이스 서버 주소
port = 3306
database = 'my_test'

# pymysql을 사용하여 MySQL 연결 엔진 생성
# 연결 문자열 포맷: DB명+driver://username:password@host:port/database
engine = create_engine(f'mysql+pymysql://{username}:{password}@{host}:{port}/{database}')

### 1. DataFrame to DB

In [18]:
# DataFrame을 MySQL 데이터베이스에 저장
# 'hospital_alchemy'은 저장하려는 테이블 이름. 이 테이블 이름이 데이터베이스에 없으면 새로 생성됨.
# if_exists 옵션은 테이블이 이미 존재할 경우 어떻게 할지 결정. 'fail', 'replace', 'append' 중 선택 가능.
df.to_sql('hospital_alchemy', con=engine, index=False, if_exists='replace')

10

### 2. DB to DataFrame

In [19]:
# SQL 쿼리 (여기서는 예제로 테이블 전체를 선택)
query = "SELECT * FROM hospital_alchemy"

# SQL 쿼리 실행 및 결과를 DataFrame으로 로드
df = pd.read_sql(query, engine)

# 결과 확인
print(df)

                                                addr clCdNm  crtrYm  \
0                        인천광역시 계양구 계양대로 85, 2층 (작전동)    한의원  202401   
1          서울특별시 구로구 경인로15길 4, SC메디칼센터 4층 401호 (오류동)   치과의원  202401   
2                      경상북도 구미시 금오대로 307-3, 1층 (오태동)    한의원  202401   
3                    대전광역시 동구 동서대로1705번길 5, 2층 (용전동)    한의원  202401   
4  서울특별시 중랑구 용마산로 670, B동 2층 206호 (망우동, 신내역 프라디움 ...    한의원  202401   
5             경기도 부천시 원미구 수도로 202, 은성프라자 2층일부호 (도당동)     의원  202401   
6             경기도 안양시 동안구 경수대로 555, 개성하이뷰 207호 (호계동)     의원  202401   
7                         강원특별자치도 원주시 북원로 2245 (단계동)     의원  202401   
8                      경기도 하남시 감일백제로 181, 403호 (감이동)    한의원  202401   
9                             전라남도 담양군 금성면 덕성길 11-58     의원  202401   

  estbCnclTp    estbDd          telno                yadmNm  \
0         개업  20240103   032-507-1075                조각목한의원   
1         개업  20240103    02-532-3423                 길치과의원   
2         개업  20240124   054-

## (번외) API와 반복문의 만남

In [20]:
url = 'http://apis.data.go.kr/B551182/yadmOpCloInfoService1/getHospPharmacyOpCloList'

df = pd.DataFrame()
for my_month in ['202301', '202302', '202303']:
    params ={'serviceKey' : serviceKey, 'numOfRows' : '10', 'pageNo' : '1', 'crtrYm' : f'{my_month}', 'yadmTp' : '1', 'opCloTp' : '1' }
    print (params)
    response = requests.get(url, params=params)
    temp_df = pd.read_xml(response.content, xpath=".//item")  # 추가된 코드
    if df.empty:
        df = temp_df
    else:
        df = pd.concat([df, temp_df], ignore_index=True)

{'serviceKey': 'gycKUW/FpEjLyAazQk7opnlSyQo44go3XD9wyf+cW165Kxw5CPTAiMbjqvd4oE7NC9mChhSIRRITlvhicq0KdQ==', 'numOfRows': '10', 'pageNo': '1', 'crtrYm': '202301', 'yadmTp': '1', 'opCloTp': '1'}
{'serviceKey': 'gycKUW/FpEjLyAazQk7opnlSyQo44go3XD9wyf+cW165Kxw5CPTAiMbjqvd4oE7NC9mChhSIRRITlvhicq0KdQ==', 'numOfRows': '10', 'pageNo': '1', 'crtrYm': '202302', 'yadmTp': '1', 'opCloTp': '1'}
{'serviceKey': 'gycKUW/FpEjLyAazQk7opnlSyQo44go3XD9wyf+cW165Kxw5CPTAiMbjqvd4oE7NC9mChhSIRRITlvhicq0KdQ==', 'numOfRows': '10', 'pageNo': '1', 'crtrYm': '202303', 'yadmTp': '1', 'opCloTp': '1'}


In [21]:
df

addr clCdNm      cnclDd  \
0             경기도 화성시 동탄대로1길 3, MJ스퀘어  401~403호 (장지동)     의원  20231130.0   
1                       인천광역시 중구 신도시북로 89, 2,3층 (운서동)     의원  20230410.0   
2                       경상남도 김해시 삼계중앙로 42, 802호 (삼계동)     의원  20230403.0   
3                                        전라북도 군산시 월명로   치과의원  20230928.0   
4                충청남도 천안시 서북구 직산읍 직산로 65, 2층 201호, 3층     병원  20240207.0   
5                        전라남도 목포시 연산로 131, 1,2층 (연산동)     의원  20240206.0   
6                                        전라북도 군산시 월명로   치과의원  20230601.0   
7             경기도 화성시 동탄대로1길 3, MJ스퀘어  401~403호 (장지동)     의원  20231023.0   
8                   강원특별자치도 강릉시 경강로2356번길 2, 2층 (포남동)    한의원  20231211.0   
9                                        전라남도 순천시 환선로     병원  20230601.0   
10                   부산광역시 동래구 충렬대로137번길 4, 10층 (온천동)    한의원  20231110.0   
11   경기도 안산시 단원구 고잔로 108, 대동조이월드 627,628,628-1호 (고잔동)   치과의원  20230504.0   
12                서울특별시 종로구 자하문로28가길 3, 지하1층,1층 (청운동)     의원  20230605.0   
13                                      대구광역시 달서구 당산로     의원  20230314.0   
14  서울특별시 강서구 양천로 570, NH서울축산농협 NH서울타워 201~203호 (등촌동)     의원  20231130.0   
15           광주광역시 서구 시청로 67, 프리미엄아울렛 1층 19-20호 (치평동)     의원  20230615.0   
16                  경상북도 구미시 구미중앙로 111, 1층 101호 (원평동)     의원  20230502.0   
17  경기도 고양시 덕양구 지축로 74, 벨라미 301호~318호,401호~418호,50...   한방병원  20230510.0   
18              서울특별시 강서구 곰달래로53가길 3, 우진타워빌딩 2층 (화곡동)    한의원  20231228.0   
19                         부산광역시 영도구 청학로 21, 1층 (청학동)    한의원  20230530.0   
20                             경상남도 창원시 의창구 창원대로363번길     의원         NaN   
21               경기도 화성시 동탄영천로 81-5, 골든프라자 302호 (영천동)     의원         NaN   
22             경기도 김포시 고촌읍 장차로13번길 11, 서익빌딩 5,3층 301호     의원         NaN   
23        경상남도 창원시 진해구 냉천로 87, Y-Zone 202호,302호호 (석동)     의원         NaN   
24                 충청남도 천안시 동남구 충절로 27, 2동 1,2층 (신부동)     의원         NaN   
25                     경상남도 창원시 의창구 동읍 용정길46번길 43, 1층    한의원         NaN   
26            부산광역시 해운대구 중동2로 16, 해운대온천센터 2층 일부호 (중동)     의원         NaN   
27                        부산광역시 수영구 연수로 399, 3층 (수영동)    한의원         NaN   
28                        대전광역시 대덕구 중리북로 47, 2층 (중리동)    한의원         NaN   
29                                      서울특별시 강남구 학동로   치과의원         NaN   

    crtrYm estbCnclTp    estbDd shwSbjtCdNm          telno       yadmNm  \
0   202301         개업  20230103         일반의  031-8055-8920      동탄스마일의원   
1   202301         개업  20230130         일반의   032-751-8577       영종하늘의원   
2   202301         개업  20230110         일반의   055-325-8370      참좋은행복의원   
3   202301         개업  20230103        None   063-461-7480       유디치과의원   
4   202301         개업  20230106        None   041-582-2200        직산큰병원   
5   202301         개업  20230119         일반의   061-276-7582        메디미의원   
6   202301         개업  20230103        None   063-461-7480       유디치과의원   
7   202301         개업  20230103         일반의  031-8055-8920     365스마일의원   
8   202301         개업  20230112        None   033-652-8875     강릉애아토한의원   
9   202301         개업  20230112        None   061-905-7588         청담병원   
10  202302         개업  20230221        None   051-501-7572        상현한의원   
11  202302         개업  20230221        None   031-411-2882     안산연세치과의원   
12  202302         개업  20230223        산부인과    02-516-7023    강마리산부인과의원   
13  202302         개업  20230214         일반의   053-557-9191        새명신의원   
14  202302         개업  20230216       가정의학과   02-2013-5688  봄앤아이가정의학과의원   
15  202302         개업  20230215         일반의   062-457-7575        더고운의원   
16  202302         개업  20230203         일반의   054-457-5421        바로그의원   
17  202302         개업  20230221        None    02-371-3650    365바로한방병원   
18  202302         개업  20230207        None   02-2646-4860      서울대찬한의원   
19  202302         개업  20230224        None   051-414-7555       한가족한의원   
20  202303         개업  20230330         일반의   055-286-8675        아너스의원   
21  202303         개업  20230324         일반